In [1]:
import numpy as np
import pandas as pd

## 分析目标

此项数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。<br>
本实战项目的目的在于联系评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据，该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：<br><br>
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以数字 "c" 开头，表示这笔交易被取消。<br>
- `StokeCode`: 产品代码。5位数，作为产品的唯一标识符。<br>
- `Description`: 产品名称。<br>
- `Quantity`: 产品在交易中的数量。<br>
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。<br>
- `UnitPrice`: 单价。价格单位为英镑(£)。<br>
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。<br>
- `Country`: 国家名称。客户做居住的国家的名称。<br>

## 读取数据

导入数据分析所需要的库，并通过 Pandas 的`read_csv`函数，将原始数据文件 "UK_e_commerce.csv" 里的数据内容，解析为 DataFrame ，并赋值给变量`Original_Data`。

In [2]:
Original_Data = pd.read_csv('./Data_Analysis_Practice//UK_e_commerce.csv', encoding='latin-1')

In [3]:
Original_Data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

在这一部分，我将对在上一部分建立的`Original_Data`这个 DataFrame 所包含的数据进行评估。<br><br>
评估主要从两个方面进行：即整齐度和干净度。数据的结构性问题指不符合"每个变量为一列，每个观察值为一行，每种类型的观察单位为一个表格"这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

## 评估数据整齐度

In [4]:
Original_Data.sample(10)

InvoiceNo StockCode                          Description  Quantity  \
401455    571441     21746         SMALL RED RETROSPOT WINDMILL         2   
265400    560225    85129D      BEADED CRYSTAL HEART PINK SMALL         2   
39408     539642     22565  FELTCRAFT HAIRBANDS PINK AND WHITE          3   
381495    569866     22759         SET OF 3 NOTEBOOKS IN PARCEL         3   
29932     538832     21871                  SAVE THE PLANET MUG        12   
524200    580537     23312          VINTAGE CHRISTMAS GIFT SACK         2   
480994    577331     23458             DOLLY CABINET 3 DRAWERS          2   
247096    558754     23209            LUNCH BAG DOILEY PATTERN          5   
42083     539958     22378                 WALL TIDY RETROSPOT         17   
85747     543518     22088          PAPER BUNTING COLOURED LACE        80   

             InvoiceDate  UnitPrice  CustomerID         Country  
401455  10/17/2011 13:31       2.46     14096.0  United Kingdom  
265400   7/15/2011 16:27       2.46         NaN  United Kingdom  
39408   12/20/2010 15:35       0.85     15867.0  United Kingdom  
381495   10/6/2011 14:50       1.65     12757.0        Portugal  
29932   12/14/2010 13:03       1.25     17158.0  United Kingdom  
524200   12/4/2011 16:08       4.15         NaN  United Kingdom  
480994  11/18/2011 14:23      14.95     14667.0  United Kingdom  
247096    7/3/2011 12:10       1.65     15518.0  United Kingdom  
42083   12/23/2010 13:26       1.66         NaN  United Kingdom  
85747     2/9/2011 11:21       2.55     12753.0           Japan

从抽样的10行数据来看，数据符合"每列是一个变量，每行是一个观察值，每个单元格是一个值"，具体来看每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。

### 评估数据干净度

In [5]:
Original_Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有 541909 条观察值，而`Description`、`CustomerID`变量存在缺失值。<br><br>
此外，`InvoiceDate`的数据类型应为日期，`CustomerID`的数据类型应为字符串，应当进行数据格式转换。

## 评估缺失数据

在了解`Description`存在缺失值后，根据条件提取出缺失观察值。

In [6]:
Original_Data[Original_Data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

有 1454 条交易数据缺失`Description`变量值。<br><br>
从输出结果看，这些缺失`Description`的交易数据，`UnitPrice`都为 0 。为了验证猜想，我们增加筛选条件，看是否存在`Description`变量缺失且`UnitPrice`不为 0 的数据。

In [7]:
Original_Data[(Original_Data["Description"].isnull()) & (Original_Data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选出来结果数量为 0 条，说明缺失`Description`值的数据，同时也不具备有效的`UnitPrice`值。<br>
`Description`表示产品名称，`UnitPrice`表示产品单价，都是进行后续产品交易分析的重要变量。如果它们同时缺失/无效，我们认为数据无法提供有效含义，因此这些后续可以被删除。

`CustomerID`变量同样存在缺失值，因此也根据条件提取出缺失观察值。

In [8]:
Original_Data[Original_Data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

`CustomerID`表示客户编号，不是分析畅销商品的必要变量。并且从输出结果来看，有些`CustomerID`确实的交易数据仍然有效，因此保留此变量为空的观察值。

#### 评估重复数据

根据变量的含义来看，虽然`InvoiceNo`、`StockCode`和`CustomerID`都是唯一标识符，但一次交易可能包含多种商品，因此`InvoiceNo`可以存在重复，不同交易可以包含同件商品，因此`StoceCode`可以存在重复，顾客可以进行多次交易或下单多个商品，因此`CustomerID`也可以存在重复。<br><br>
那么针对此数据集，我们无需评估重复数据。

#### 评估不一致数据

不一致数据可能存在于`Country`变量中，我们要查看是否存在多个不同值指代同一国家的情况。

In [9]:
Original_Data["Country"].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


这里可以看出，并无重复的国家和地区名在`Country`栏中。但视频中此处讲到将出现重复名的国家内容部分重新整理，因此在下列内容中，按照出现重复国家名的情况进行操作。

↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓

从`Country`变量值来看，'"USA"'、'"United States"'均在表示美国，'"United Kindom"'、'"UK"'、'"U.K."'均在表示英国，因此应该在清洗步骤对这些值进行统一，只保留一个指代值。

↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑

In [10]:
Original_Data01 = pd.read_csv('./Data_Analysis_Practice//e_commerce_UK.csv', encoding='latin-1')

In [11]:
Original_Data01["Country"].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


我后来又找到了一份7年前的数据(第一份数据是一年前发布上去的)，可以看出还是没有出现不同名称指向同一国家的问题，因此先这么着吧。

### 评估无效或错误数据

In [12]:
Original_Data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，`Quantity`和`UnitPrice`存在负数，会对后续数值分析产生影响。<br>
因此，我们先筛选出`Quantity`数值为负数的观察值，进一步评估其含义。

In [13]:
Original_Data[Original_Data['Quantity'] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看，`Quantity`变量为负的观察值，`InvoiceNo`似乎均以 C 开头，表示订单被取消。<br>
但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在`Quantity`变量为负数且`InvoiceNo`不以 C 开头的观察值。

In [14]:
Original_Data[(Original_Data['Quantity'] < 0) & (Original_Data['InvoiceNo'].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被证实错误，因为还存在`Quantity`变量为负数且`InvoiceNo`不以 C 开头的观察值。<br>
但根据以上输出结果，这些筛选出的观察值的`UnitPrice`观察值均为 0 , 因此增加`UnitPrice`的条件进行验证。

In [15]:
Original_Data[(Original_Data['Quantity'] < 0) 
              & (Original_Data['InvoiceNo'].str[0] != "C")
              & (Original_Data['UnitPrice'] != 0) ]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，当`Quantity`变量为负数的时候，观察值满足以下条件之一:
> - `InVoiceNO`以 C 开头，表示订单取消<br>
> - `UnitPrice`为 0 , 说明单价为 0 英镑<br>

这些交易数据均不是有效成交数据不贡献销售不在后续分析范围内，因此我们在数据清理步骤将`Quantity`为负数的观察值删除。

In [16]:
Original_Data[Original_Data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，`UnitPrice`为负数的观察值都是坏账调整，不属于实际商品交易数据，因此也在数据清理步骤中也将其删除。

## 清理数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括:
- 把`InvoiceDate`变量的数据类型转换为日期时间
- 把`CustomerID`变量的数据类型转换为字符串
- 把`Description`变量确实的观察值删除
- 把`Country`变量值`USA`替换为`United States` (本数据表中没有重复的国家或地区名，但保留视频中的操作过程)
- 把`Country`变量值`UK`、`U.K.`替换为`United Kindom` (本数据表中没有重复的国家或地区名，但保留视频中的操作过程)
- 把`Quantity`变量值为负数的观察值删除
- 把`UnitPrice`变量值为负数的观察值删除

为了区分经过清理的数据和原始的数据，我们创建新的变量`Cleaned_Data`，让它为`Original_Data`复制出的副本，我们之后的清理步骤都将被运用在`Cleaned_Data`上。

In [17]:
Cleaned_Data = Original_Data.copy()
Cleaned_Data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

把`InvoiceDate`变量的数据类型转换为日期时间类型:

In [22]:
Cleaned_Data["InvoiceDate"] = pd.to_datetime(Cleaned_Data["InvoiceDate"])
Cleaned_Data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把`CustomerID`变量的数据类型转换为字符串

In [23]:
Cleaned_Data["CustomerID"] = Cleaned_Data["CustomerID"].astype(str)
Cleaned_Data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把`CustomerID`变量值结尾的`".0"`删除:

In [25]:
Cleaned_Data["CustomerID"] =Cleaned_Data["CustomerID"].str.slice(0,-2)
Cleaned_Data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把`Description`变量确实的观察值删除，并查看删除后该列空缺值个数和：

In [26]:
Cleaned_Data.dropna(subset=["Description"])

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85      12680          France  
541905 2011-12-09 12:50:00       2.10      12680          France  
541906 2011-12-09 12:50:00       4.15      12680          France  
541907 2011-12-09 12:50:00       4.15      12680          France  
541908 2011-12-09 12:50:00       4.95      12680          France  

[540455 rows x 8 columns]

In [27]:
Cleaned_Data.dropna(subset=["Description"], inplace=True)

In [28]:
Cleaned_Data["Description"].isnull().sum()

np.int64(0)

把`Country`变量值`USA`替换为`United States` <br>
(本数据表中没有重复的国家或地区名，但保留视频中的操作过程)

In [29]:
Cleaned_Data["Country"] = Cleaned_Data["Country"].replace({"USA": "United States"})

In [30]:
len(Cleaned_Data[Cleaned_Data["Country"] == "USA"])

0

把`Country`变量值`UK`、`U.K.`替换为`United Kindom` <br>
(本数据表中没有重复的国家或地区名，但保留视频中的操作过程)

In [32]:
Cleaned_Data["Country"] = Cleaned_Data["Country"].replace({"UK": "United Kindom", "U.K.": "United Kindom"})

In [33]:
print(len(Cleaned_Data[Cleaned_Data["Country"] == "UK"]))
print(len(Cleaned_Data[Cleaned_Data["Country"] == "U.K."]))

0
0


把`Quantity`变量值为负数的观察值删除，并检查替换后`Quantity`变量值为负数的个数:

In [34]:
Cleaned_Data = Cleaned_Data[Cleaned_Data["Quantity"] >= 0]

In [36]:
len(Cleaned_Data[Cleaned_Data["Quantity"] < 0])

0

把`UnitPrice`变量值为负数的观察值删除，并检查替换后`UnitPrice`变量值为负数的个数:

In [37]:
Cleaned_Data = Cleaned_Data[Cleaned_Data["UnitPrice"] >= 0]

In [38]:
len(Cleaned_Data[Cleaned_Data["UnitPrice"] < 0])

0

## 保存清理后的数据

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为`UK_e_commerce_cleaned.csv`

In [39]:
Cleaned_Data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

In [40]:
Cleaned_Data.to_csv("./Data_Analysis_Practice//UK_e_commerce_cleaned.csv", index= False)

In [41]:
pd.read_csv("./Data_Analysis_Practice//UK_e_commerce_cleaned.csv").head()

C:\Users\Administrator\AppData\Local\Temp\ipykernel_8940\3089635424.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("./Data_Analysis_Practice//UK_e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom